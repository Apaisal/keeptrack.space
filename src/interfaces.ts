import { BaseObject, Degrees, DetailedSatellite, DetailedSensor, GreenwichMeanSiderealTime, Kilometers, Meters, Milliseconds, Radians, SpaceObjectType, Vec3Flat } from 'ootk';
import { Camera } from './singletons/camera';
import { GroupType, ObjectGroup } from './singletons/object-group';

import { mat4 } from 'gl-matrix';
import { ZoomValue } from 'ootk/lib/objects/DetailedSensor';
import { MissileObject } from './singletons/catalog-manager/MissileObject';
import { ColorSchemeManager } from './singletons/color-scheme-manager';
import { LineManager } from './singletons/draw-manager/line-manager';
import { HoverManager } from './singletons/hover-manager';
import { SearchManager } from './singletons/search-manager';
import { TearrData } from './static/sensor-math';

/** Array of ECI Coordinates [x, y, z] */
export type EciArr3 = Vec3Flat<number>;

/** Array of ECF Coordinates [x, y, z] */
export type EcfArr3 = Vec3Flat<number>;

export type Constructor<I> = new (...args: any[]) => I;

export interface SatShader {
  blurFactor1: string;
  blurFactor2: string;
  blurFactor3: string;
  blurFactor4: string;
  distanceBeforeGrow: string;
  dynamicSizeScalar: number;
  isUseDynamicSizing: boolean;
  largeObjectMaxZoom: number;
  largeObjectMinZoom: number;
  maxAllowedSize: number;
  maxSizePlanetarium: number;
  minSizePlanetarium: number;
  starSize: string;
}

export enum GetSatType {
  DEFAULT = 0,
  EXTRA_ONLY = 1,
  POSITION_ONLY = 2,
  SKIP_POS_VEL = 3,
}

export type MissileParams = {
  desc: string;
  active: boolean;
  latList: Degrees[];
  lonList: Degrees[];
  altList: Kilometers[];
  timeList: number[];
  startTime: number;
  maxAlt: number;
  country: string;
  launchVehicle: string;
};

export type EChartsData = {
  name: string;
  satId?: number;
  country?: string;
  value?: [number, number, number][];
  data?: [number, number, number][];
}[];

export type rgbaArray = [number, number, number, number];

export interface Colors {
  notional: rgbaArray;
  ageLost: rgbaArray;
  ageMed: rgbaArray;
  ageNew: rgbaArray;
  ageOld: rgbaArray;
  analyst: rgbaArray;
  countryCIS: rgbaArray;
  countryOther: rgbaArray;
  countryPRC: rgbaArray;
  countryUS: rgbaArray;
  debris: rgbaArray;
  densityHi: rgbaArray;
  densityLow: rgbaArray;
  densityMed: rgbaArray;
  densityOther: rgbaArray;
  densityPayload: rgbaArray;
  deselected: rgbaArray;
  facility: rgbaArray;
  gradientAmt: number;
  inFOV: rgbaArray;
  inFOVAlt: rgbaArray;
  inGroup: rgbaArray;
  length: number;
  lostobjects: rgbaArray;
  marker: rgbaArray[];
  missile: rgbaArray;
  missileInview: rgbaArray;
  payload: rgbaArray;
  penumbral: rgbaArray;
  pink: rgbaArray;
  confidenceHi: rgbaArray;
  confidenceLow: rgbaArray;
  confidenceMed: rgbaArray;
  rcsLarge: rgbaArray;
  rcsMed: rgbaArray;
  rcsSmall: rgbaArray;
  rcsUnknown: rgbaArray;
  rcsXSmall: rgbaArray;
  rcsXXSmall: rgbaArray;
  rocketBody: rgbaArray;
  satGEO: rgbaArray;
  satHi: rgbaArray;
  satLEO: rgbaArray;
  satLow: rgbaArray;
  satMed: rgbaArray;
  satSmall: rgbaArray;
  sensor: rgbaArray;
  starHi: rgbaArray;
  starLow: rgbaArray;
  starMed: rgbaArray;
  sunlight100: rgbaArray;
  sunlight60: rgbaArray;
  sunlight80: rgbaArray;
  sunlightInview: rgbaArray;
  transparent: rgbaArray;
  trusat: rgbaArray;
  umbral: rgbaArray;
  unknown: rgbaArray;
  version: string;
}

export interface SatCruncherMessageData {
  data: {
    typ?: string;
    time?: number;
    propRate?: number;
    staticOffset?: number;
    dynamicOffsetEpoch?: number;
    badSatNumber?: number;
  };
}

export interface SensorObjectCruncher {
  alt: Kilometers;
  beamwidth?: Degrees;
  changeObjectInterval?: Milliseconds;
  country: string;
  lat: Degrees;
  linkAehf?: boolean;
  linkWgs?: boolean;
  lon: Degrees;
  name: string;
  observerGd: {
    lat: Radians;
    lon: Radians;
    alt: Kilometers;
  };
  maxAz: Degrees;
  maxAz2?: Degrees;
  maxEl: Degrees;
  maxEl2?: Degrees;
  maxRng: Kilometers;
  maxRng2?: Kilometers;
  minAz: Degrees;
  minAz2?: Degrees;
  minEl: Degrees;
  minEl2?: Degrees;
  minRng: Kilometers;
  minRng2?: Kilometers;
  shortName: string;
  sensorId: number;
  sun: string;
  type?: SpaceObjectType;
  url?: string;
  volume: boolean;
  zoom: ZoomValue;
  objName: string;
  uiName: string;
  system: string;
  operator: string;
}

export type lookanglesRow = {
  sortTime: number;
  scc: string;
  score: number;
  startDate: Date;
  startTime: Date;
  startAz: string;
  startEl: string;
  startrng: string;
  stopDate: Date;
  stopTime: Date;
  stopAz: string;
  stopEl: string;
  stoprng: string;
  tic: number;
  minrng: string;
  passMaxEl: string;
};

export type SatPassTimes = {
  sat: DetailedSatellite;
  time: Date;
};

export type ToastMsgType = 'standby' | 'normal' | 'caution' | 'serious' | 'critical' | 'error';

export type GeolocationPosition = {
  coords: {
    latitude: Degrees | number;
    longitude: Degrees | number;
    altitude: Meters | number;
  };
};

export interface SensorGeolocation {
  alt: Kilometers;
  lat: Degrees;
  lon: Degrees;
  maxaz: Degrees;
  maxel: Degrees;
  maxrange: Kilometers;
  minaz: Degrees;
  minel: Degrees;
  minrange: Kilometers;
}

export interface ObjectTypeFlags {
  ageLost: boolean;
  ageMed: boolean;
  ageNew: boolean;
  ageOld: boolean;
  countryCIS: boolean;
  countryOther: boolean;
  countryPRC: boolean;
  countryUS: boolean;
  debris: boolean;
  densityHi: boolean;
  densityLow: boolean;
  densityMed: boolean;
  densityOther: boolean;
  densityPayload: boolean;
  facility: boolean;
  inFOV: boolean;
  inViewAlt: boolean;
  missile: boolean;
  missileInview: boolean;
  payload: boolean;
  pink: boolean;
  rcsLarge: boolean;
  rcsMed: boolean;
  rcsSmall: boolean;
  rcsUnknown: boolean;
  rocketBody: boolean;
  satGEO: boolean;
  satHi: boolean;
  satLEO: boolean;
  satLow: boolean;
  satMed: boolean;
  satSmall: boolean;
  sensor: boolean;
  starHi: boolean;
  starLow: boolean;
  starMed: boolean;
  velocityFast: boolean;
  velocityMed: boolean;
  velocitySlow: boolean;
}

export enum Pickable {
  Yes = 1,
  No = 0,
}

export type ColorInformation = {
  color: [number, number, number, number];
  marker?: boolean;
  pickable: Pickable;
};

export interface ColorRuleParams {
  satInSun?: Int8Array;
  satInView?: Int8Array;
}

export type ColorRuleSet = (obj: BaseObject, params?: any) => ColorInformation;

/************************************************************
 * Singletons
 *
 * Using these interfaces will allow us to create replaceable
 * singletons. In the future, we can create new classes that
 * implement these interfaces and replace the singletons with
 * the new classes. Example: A new class that implements the
 * CatalogManager interface but is more resource intensive
 * for better accuracy.
 *
 * **********************************************************/

export interface GroupsManager {
  clearSelect: () => void;
  createGroup: (type: GroupType, data: any, name?: string) => ObjectGroup;
  groupList: Record<string, ObjectGroup>;
  selectGroup: (group: ObjectGroup) => void;
  selectedGroup: ObjectGroup;
  stopUpdatingInViewSoon: boolean;
}

export interface OrbitManager {
  orbitWorker: Worker;
  tempTransColor: [number, number, number, number];

  addInViewOrbit(satId: number): void;
  changeOrbitBufferData(satId: number, TLE1?: string, TLE2?: string): void;
  clearHoverOrbit(): void;
  clearInViewOrbit(): void;
  drawOrbitsSettingChanged(): void;
  clearSelectOrbit(isSecondary?: boolean): void;
  draw(
    pMatrix: mat4,
    camMatrix: mat4,
    tgtBuffer: WebGLFramebuffer,
    hoverManagerInstance: HoverManager,
    colorSchemeManagerInstance: ColorSchemeManager,
    mainCameraInstance: Camera
  ): void;
  init(lineManagerInstance: LineManager, gl: WebGL2RenderingContext, orbitWorker?: Worker): void;
  removeInViewOrbit(satId: number): void;
  setHoverOrbit(satId: number): void;
  setSelectOrbit(satId: number, isSecondary?: boolean): void;
  updateAllVisibleOrbits(uiManagerInstance: UiManager): void;
  updateOrbitBuffer(
    satId: number,
    missileParams?: {
      latList: Degrees[];
      lonList: Degrees[];
      altList: Kilometers[];
    }
  ): void;
}

export interface SensorManager {
  lastMultiSiteArray: TearrData[];
  currentSensors: DetailedSensor[];
  customSensors: DetailedSensor[];
  defaultSensor: DetailedSensor[];
  isCustomSensorMenuOpen: boolean;
  isLookanglesMenuOpen: boolean;
  secondarySensors: DetailedSensor[];
  sensorListUS: DetailedSensor[];
  sensorTitle: string;
  stfSensors: DetailedSensor[];
  /** Deprecated - Stop using this */
  whichRadar: string;

  addSecondarySensor(sensor: DetailedSensor): void;
  addStf(sensor: DetailedSensor): void;
  clearSecondarySensors(): void;
  clearStf(): void;
  isSensorSelected(): boolean;
  removeSecondarySensor(sensor: DetailedSensor): void;
  removeStf(sensor?: DetailedSensor): void;
  resetSensorSelected(): void;
  setCurrentSensor(sensor: DetailedSensor[] | null): void;
  setSensor(selectedSensor: DetailedSensor | string | null, sensorId?: number): void;
  updateCruncherOnCustomSensors(): void;
  verifySensors(sensors: DetailedSensor[]): DetailedSensor[];
  calculateSensorPos(now: Date, sensors?: DetailedSensor[]): { x: number; y: number; z: number; lat: number; lon: number; gmst: GreenwichMeanSiderealTime };
}

export interface UiManager {
  bottomIconPress: (el: HTMLElement) => void;
  hideSideMenus: () => void;
  dismissAllToasts: () => void;
  isAnalysisMenuOpen: boolean;
  isCurrentlyTyping: boolean;
  isUiVisible: boolean;
  lastBoxUpdateTime: number;
  lastColorScheme: ColorRuleSet;
  lastNextPassCalcSatId: number;
  lastNextPassCalcSensorShortName: string;
  lastToast: string;
  lookAtLatLon: undefined;
  searchManager: SearchManager;
  updateInterval: number;
  updateNextPassOverlay: (boolean) => void;
  searchHoverSatId: number;
  colorSchemeChangeAlert: (scheme: ColorRuleSet) => void;
  doSearch(searchString: string, isPreventDropDown?: boolean): void;
  footerToggle(): void;
  hideUi(): void;
  init(): void;
  initMenuController(): void;
  legendHoverMenuClick(legendType: string): void;
  onReady(): void;
  toast(toastText: string, type: ToastMsgType, isLong?: boolean): void;
  updateSelectBox(realTime: Milliseconds, lastBoxUpdateTime: Milliseconds, satOrMislObj: DetailedSatellite | MissileObject): void;
}

export enum Singletons {
  CatalogManager = 'CatalogManager',
  OrbitManager = 'OrbitManager',
  GroupsManager = 'GroupManager',
  UiManager = 'UiManager',
  ColorSchemeManager = 'ColorScheme',
  HoverManager = 'HoverManager',
  LineManager = 'LineManager',
  StarManager = 'StarManager',
  TimeManager = 'TimeManager',
  WebGLRenderer = 'DrawManager',
  DotsManager = 'DotsManager',
  SensorManager = 'SensorManager',
  SelectSatManager = 'SelectSatManager',
  InputManager = 'InputManager',
  SoundManager = 'SoundManager',
  SensorMath = 'SensorMathManager',
  MainCamera = 'MainCamera',
  PersistenceManager = 'PersistenceManager',
  Scene = 'Scene',
  MeshManager = 'MeshManager',
}
/**
 * Enum containing the registrable events used in the KeepTrack API.
 */

export enum KeepTrackApiEvents {
  onHelpMenuClick = 'onHelpMenuClick',
  /**
   * Run at the end of SelectSatManager.selectSat with parameters (sat: SatObject, satId: number)
   */
  selectSatData = 'selectSatData',
  /**
   * Run at the end of catalogManager.setSecondarySat with parameters (sat: SatObject, satId: number)
   */
  setSecondarySat = 'setSecondarySat',
  onKeepTrackReady = 'onKeepTrackReady',
  updateSelectBox = 'updateSelectBox',
  onCruncherReady = 'onCruncherReady',
  onCruncherMessage = 'onCruncherMessage',
  uiManagerInit = 'uiManagerInit',
  uiManagerOnReady = 'uiManagerOnReady',
  bottomMenuClick = 'bottomMenuClick',
  hideSideMenus = 'hideSideMenus',
  nightToggle = 'nightToggle',
  orbitManagerInit = 'orbitManagerInit',
  drawManagerLoadScene = 'drawManagerLoadScene',
  drawOptionalScenery = 'drawOptionalScenery',
  updateLoop = 'updateLoop',
  /**
   * Run as the default case in the rmbMenuActions event with parameters (targetId: string, clickedSat: number)
   */
  rmbMenuActions = 'rmbMenuActions',
  /**
   * Runs during inputManager.init immediately before adding the clear lines and clear screen buttons
   */
  rightBtnMenuAdd = 'rightBtnMenuAdd',
  updateDateTime = 'updateDateTime',
  uiManagerFinal = 'uiManagerFinal',
  resetSensor = 'resetSensor',
  /**
   * Run in the setSensor method of SensorManager instance with parameters (sensor: DetailedSensor | string, staticId: number)
   */
  setSensor = 'setSensor',
  changeSensorMarkers = 'changeSensorMarkers',
  altCanvasResize = 'altCanvasResize',
  endOfDraw = 'endOfDraw',
  /**
   * Run in the updateWatchlist method of CatalogManager instance with parameters (watchlist: number[])
   */
  onWatchlistUpdated = 'onWatchlistUpdated',
  /**
   * Run in the staticOffset setter of TimeManager instance with parameters (staticOffset: number)
   */
  staticOffsetChange = 'staticOffsetChange',
  /**
   * Runs when a line is added to the line manager
   */
  onLineAdded = 'onLineAdded',
  /**
   * Runs when a sensor dot is selected but not when a sensor is selected from the sensor menu
   */
  sensorDotSelected = 'sensorDotSelected',
  canvasMouseDown = 'canvasMouseDown',
  touchStart = 'touchStart',
}
