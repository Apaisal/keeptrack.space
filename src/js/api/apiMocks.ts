import { satObject, sensorObject } from './keepTrack';

export const defaultSat: satObject = {
  active: true,
  SCC_NUM: '00005',
  C: 'ISS',
  LS: 'TTMTR',
  LV: 'Proton-K',
  ON: 'ISS (ZARYA)',
  OT: 1,
  TLE1: '1 25544U 98067A   21203.40407588  .00003453  00000-0  71172-4 0  9991',
  TLE2: '2 25544  51.6423 168.5744 0001475 184.3976 313.3642 15.48839820294053',
  R: '99.0524',
  URL: 'http://www.esa.int/export/esaHS/ESAI2X0VMOC_iss_0.html',
  O: 'National Aeronautics and Space Administration (NASA)/Multinational',
  U: 'Government',
  P: 'Space Science',
  LM: '',
  DM: '',
  Pw: '',
  Li: 30,
  Con: 'Boeing Satellite Systems (prime)/Multinational',
  M: 'Final size of a Boeing 747; first component.',
  S1: 'http://www.esa.int/export/esaHS/ESAI2X0VMOC_iss_0.html',
  S2: '',
  S3: 'http://www.boeing.com/defense-space/space/spacestation/overview/',
  S4: '',
  S5: '',
  S6: '',
  S7: '',
  inclination: 51.6423,
  longitude: 168.5744,
  perigee: 1475,
  apogee: 184.3976,
  period: 313.3642,
  meanMotion: 15.48839820294053,
  semimajorAxis: 21203.40407588,
  eccentricity: 0.00003453,
  raan: 0,
  argPe: 0,
  inview: 1,
  velocity: {
    total: 0,
    x: 0,
    y: 0,
    z: 0,
  },
  getTEARR: () => ({
    lon: 168.5744,
    lat: 51.6423,
    alt: 1475,
  }),
  getAltitude: () => 0,
  getDirection: () => 'N',
};

export const defaultSensor: sensorObject = {
  alt: 0.060966,
  beamwidth: 2,
  changeObjectInterval: 1000,
  country: 'United States',
  lat: 41.754785,
  linkAehf: true,
  linkWgs: true,
  lon: -70.539151,
  name: 'Cape Cod AFS, Massachusetts',
  observerGd: { lat: 0.7287584767123405, lon: -1.2311404365114507, alt: 0.060966 },
  obsmaxaz: 227,
  obsmaxel: 85,
  obsmaxrange: 5556,
  obsminaz: 347,
  obsminel: 3,
  obsminrange: 200,
  shortName: 'COD',
  staticNum: 0,
  sun: 'No Impact',
  type: 'Phased Array Radar',
  url: 'http://www.radartutorial.eu/19.kartei/01.oth/karte004.en.html',
  volume: false,
  zoom: 'leo',
};

export const useMockWorkers = () => {
  class Worker {
    url: any;
    onmessage: (msg: any) => void;
    constructor(stringUrl: any) {
      this.url = stringUrl;
      this.onmessage = () => {
        return {
          data: {}
        }
      };
    }

    postMessage(msg: any) {
      if (msg.dat) {
        this.onmessage({          
          data: {
            extraData: JSON.stringify([defaultSat, defaultSensor]),
            satPos: [0,0,0,0,0,0],
            satVel: [0,0,0,0,0,0],
          }
        });
      }
      this.onmessage({
        data: {
          satPos: [0,0,0,0,0,0],
          satVel: [0,0,0,0,0,0],
        }
      });
    }
  }

  (<any>window).Worker = Worker;
}

export const keepTrackApiStubs = {
  programs: {
    adviceManager: {
      adviceList: {
        satViewDisabled: jest.fn(),
        editSatDisabled: jest.fn(),
        survFenceDisabled: jest.fn(),
        bubbleDisabled: jest.fn(),
        satFOVDisabled: jest.fn(),
        planetariumDisabled: jest.fn(),
        mapDisabled: jest.fn(),
      },
      adviceCount: {
        socrates: 0,
      },
      showAdvice: jest.fn(),
      adviceArray: (<any[]>[]),
    },
    mainCamera: {
      cameraType: {
        current: 1,
        satellite: 1,
        fixedToSat: 2,
      },
      camSnap: jest.fn(),
      changeZoom: jest.fn(),
      latToPitch: jest.fn(),
      longToYaw: jest.fn(),
      autoRotate: jest.fn(),
    },
    ColorScheme: {
      reloadColors: jest.fn(),      
    },
    drawManager: {
      glInit: jest.fn(),
      gl: global.mocks.glMock,
      selectSatManager: {
        selectSat: jest.fn(),
      },
    },    
    dotsManager: {
      starIndex1: 0,
      pickReadPixelBuffer: [
        0,
        0,
        0,
      ],
    },    
    groups: null,
    groupsManager: {
      Canada: null,
      Japan: null,
      France: null,
      China: null,
      India: null,
      Israel: null,
      Russia: null,
      UnitedKingdom: null,
      UnitedStates: null,
      SpaceStations: {
        sats: [
          {satId: 1, SCC_NUM: '25544'},
          {satId: 1, SCC_NUM: '25544'},
        ],
        updateOrbits: jest.fn(),
      },
      clearSelect: jest.fn(),
      createGroup: () => ({
        updateOrbits: jest.fn(),
      }),
      selectGroup: jest.fn(),
    },    
    lineManager: {
      create: jest.fn(),
      removeStars: jest.fn(),
      getLineListLen: jest.fn(),
    },
    missileManager: {
      getMissileTEARR: jest.fn(),
    },
    objectManager: {
      isLaunchSiteManagerLoaded: true,
      launchSiteManager: {
        launchSiteList: [
          'AFWTR'
        ],
      },
      selectedSat: -1,
      setSelectedSat: jest.fn(),
      staticSet: {
        id: 0,
      },
      analSatSet: {
        id: 0,
      },
      radarDataSet: {
        id: 0,
      },
      missileSet: {
        id: 0,
      },
      fieldOfViewSet: {
        id: 0,
      },
    },
    orbitManager: {
      updateOrbitBuffer: jest.fn(),
      clearInViewOrbit: jest.fn(),
      removeInViewOrbit: jest.fn(),
      historyOfSatellitesPlay: jest.fn(),
    },
    satellite: {
      currentTEARR: {
        az: 0,
        el: 0,
        rng: 0,
      },
      sgp4: () => ([0,0,0]),
      setobs: jest.fn(),
      setTEARR: jest.fn(),
      altitudeCheck: jest.fn(),
      degreesLong: (num: number) => num,
      degreesLat: (num: number) => num,
      currentEpoch: () => ['21', '150'],
      getOrbitByLatLon: () => ['1 25544U 98067A   21203.40407588  .00003453  00000-0  71172-4 0  9991', '2 25544  51.6423 168.5744 0001475 184.3976 313.3642 15.48839820294053'],
      nextpassList: () => [new Date().getTime(), new Date().getTime() + 1],
      twoline2satrec: () => ({ jdsatepoch: 2458000.5 }),
      getRae: () => [0, 0, 0],
      checkIsInFOV: () => true,
      findCloseObjects: () => '',
      findBestPasses: () => jest.fn(),
      nextpass: () => '2021-09-21 20:00:00Z',
      getDOPsTable: jest.fn(),
    },
    satSet: {
      satCruncher: {
        postMessage: jest.fn(),
      },
      satData: [
        defaultSat,
        defaultSat,
      ],
      getIdFromObjNum: jest.fn(),
      getSensorFromSensorName: () => 0,
      getSatExtraOnly: () => defaultSat,
      getSatFromObjNum : () => defaultSat,
      getSat: () => defaultSat,
      initGsData: jest.fn(),
      setColorScheme: jest.fn(),
      getSatInViewOnly: () => defaultSat,
      default: '',
      insertNewAnalystSatellite: jest.fn(),
    },
    searchBox: {
      hideResults: jest.fn(),
      fillResultBox: jest.fn(),
    },
    sensorFov: {
      enableFovView: jest.fn(),
    },
    sensorManager: {
      checkSensorSelected: () => false,
      currentSensor: {
        alt: 0,
        lat: 0,
        lon: 0,
        shortName: 'COD',
      },
    },
    starManager: {
      init: jest.fn(),
    },
    timeManager: {
      propOffset: 0,
      propRate: 0,
      selectedDate: 0,
      propTime: () => 0,
      getDayOfYear: () => 0,
    },
    uiManager: {
      hideSideMenus: jest.fn(),
      toast: jest.fn(),
      colorSchemeChangeAlert: jest.fn(),
      bottomIconPress: jest.fn(),
      legendColorsChange: jest.fn(),
      doSearch: jest.fn(),
      loadStr: jest.fn(),
      searchToggle: jest.fn(),
      legendMenuChange: jest.fn(),
      clearRMBSubMenu: jest.fn(),
      updateURL: jest.fn(),
    },
    watchlist: {
      updateWatchlist: jest.fn(),
    },
  },
};

keepTrackApiStubs.programs.groupsManager.Canada = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.Japan = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.France = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.China = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.India = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.Israel = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.Russia = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.UnitedKingdom = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.UnitedStates = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groups = keepTrackApiStubs.programs.groupsManager;