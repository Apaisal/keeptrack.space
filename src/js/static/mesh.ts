import { BufferGeometry } from './buffer-geometry';
import { GlUtils } from './gl-utils';
import { GLSL1, GLSL3, Material } from './material';

export class Mesh {
  name: string;
  gl: WebGL2RenderingContext;
  geometry: BufferGeometry;
  material: Material;
  program: WebGLProgram;

  constructor(gl: WebGL2RenderingContext, geometry: BufferGeometry, material: Material, name?: string) {
    const precision = 'highp'; // 'mediump' 'lowp'

    this.gl = gl;
    this.geometry = geometry;
    this.material = material;
    this.name = name;

    let vertexShaderHeader = '';
    let fragmentShaderHeader = '';

    if (this.material.glslVersion === GLSL3) {
      vertexShaderHeader = `#version 300 es\n`;
      fragmentShaderHeader = `#version 300 es\n`;
    }
    if (this.material.glslVersion === GLSL1) {
      vertexShaderHeader = `#version 100\n`;
      fragmentShaderHeader = `#version 100\n`;
    }

    vertexShaderHeader += `precision ${precision} float;\n\n`;
    fragmentShaderHeader += `precision ${precision} float;\n\n`;

    vertexShaderHeader += `uniform mat4 modelMatrix;\n`;
    vertexShaderHeader += `uniform mat4 modelViewMatrix;\n`;
    vertexShaderHeader += `uniform mat4 projectionMatrix;\n`;
    vertexShaderHeader += `uniform mat4 viewMatrix;\n`;
    vertexShaderHeader += `uniform mat3 normalMatrix;\n`;
    vertexShaderHeader += `uniform vec3 cameraPosition;\n`;

    if (this.material.glslVersion === GLSL3) {
      vertexShaderHeader += `
      in vec3 position;
      in vec3 normal;
      in vec2 uv;`;
    }

    if (this.material.glslVersion === GLSL1) {
      vertexShaderHeader += `
      attribute vec3 position;
      attribute vec3 normal;
      attribute vec2 uv;`;
    }

    fragmentShaderHeader += `
    uniform mat4 viewMatrix;
    uniform vec3 cameraPosition;`;

    this.material.vertexShader = `${vertexShaderHeader}${this.material.vertexShader}`;
    this.material.fragmentShader = `${fragmentShaderHeader}${this.material.fragmentShader}`;

    this.program = GlUtils.createProgram(gl, material.vertexShader, material.fragmentShader, geometry.attributes, material.uniforms, name);
    material.uniforms = Object.fromEntries(Object.entries(material.uniforms).filter(([_, v]) => v !== null));
    geometry.attributes = Object.fromEntries(Object.entries(geometry.attributes).filter(([_, v]) => v.location !== -1));
  }
}
