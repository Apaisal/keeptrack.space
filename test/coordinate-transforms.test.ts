// Generated by CodiumAI

import { EciArr3, SatObject } from '@app/interfaces';
import { CoordinateTransforms } from '@app/static/coordinate-transforms';
import { Degrees, EciVec3, Kilometers, Radians } from 'ootk';
import { defaultSensor } from './environment/apiMocks';

/**
Main functionalities:
The CoordinateTransforms class provides static methods for converting between different coordinate systems used in satellite tracking. It includes methods for converting satellite position and velocity vectors from ECI to RIC reference frame, converting RAE (Right Ascension, Azimuth, Elevation) coordinates to ECEF (Earth-Centered, Earth-Fixed) coordinates, converting ECI coordinates to latitude, longitude and altitude, and converting ECI coordinates to RAE coordinates.

Methods:
- sat2ric(sat: SatObject, reference: SatObject): Converts satellite position and velocity vectors from ECI to RIC reference frame.
- rae2ecf(az: Degrees, el: Degrees, rng: Kilometers, lat: Radians, lon: Radians, alt: Kilometers): Converts RAE coordinates to ECEF coordinates.
- eci2lla(position: EciVec3, simulationTime: Date): Converts ECI coordinates to latitude, longitude and altitude.
- eci2rae(now: Date, eci: EciArr3, sensor: SensorObject): Converts ECI coordinates to RAE coordinates.

Fields:
- None.
 */

describe('CoordinateTransforms_class', () => {
  // Tests that sat2ric method returns correct RIC coordinates for valid input
  it('test_sat2ric_valid_input', () => {
    const sat = {
      position: { x: 1000, y: 2000, z: 3000 },
      velocity: { x: 7, y: 0, z: 0 },
    };
    const reference = {
      position: { x: -1000, y: -2000, z: -3000 },
      velocity: { x: 7, y: 0, z: 0 },
    };

    const result = CoordinateTransforms.sat2ric(sat as SatObject, reference as SatObject);

    expect(result).toMatchSnapshot();
  });

  // Tests that rae2ecf method returns correct ECEF coordinates for valid input
  it('test_rae2ecf_valid_input', () => {
    const az = 45 as Degrees;
    const el = 30 as Degrees;
    const rng = 1000 as Kilometers;
    const lat = 0.7853981633974483 as Radians;
    const lon = -1.5707963267948966 as Radians;
    const alt = 0 as Kilometers;
    const result = CoordinateTransforms.rae2ecf(az, el, rng, lat, lon, alt);
    expect(result).toMatchSnapshot();
  });

  // Tests that eci2lla method returns correct LLA coordinates for valid input and simulation time
  it('test_eci2lla_valid_input', () => {
    const position = { x: -445.041, y: -445.041, z: 707.106 } as EciVec3;
    const simulationTime = new Date('2021-01-01T00:00:00Z');
    const result = CoordinateTransforms.eci2lla(position, simulationTime);

    expect(result).toMatchSnapshot();
  });

  // Tests that sat2ric method throws an error for null input
  it('test_sat2ric_null_input', () => {
    const sat = null;
    const reference = null;
    expect(() => CoordinateTransforms.sat2ric(sat, reference)).toThrow();
  });

  // Tests that rae2ecf method throws an error for negative range
  it('test_rae2ecf_negative_range', () => {
    const az = 45 as Degrees;
    const el = 30 as Degrees;
    const rng = -1000 as Kilometers;
    const lat = 0.7853981633974483 as Radians;
    const lon = -1.5707963267948966 as Radians;
    const alt = 0 as Kilometers;
    expect(() => CoordinateTransforms.rae2ecf(az, el, rng, lat, lon, alt)).toThrow();
  });

  // Tests that eci2lla method throws an error for invalid input and simulation time
  it('test_eci2lla_invalid_input', () => {
    const position = null;
    const simulationTime = null;
    expect(() => CoordinateTransforms.eci2lla(position, simulationTime)).toThrow();
  });

  it('test_valid_input_with_observerGd_and_positionEcf', () => {
    const now = new Date();
    const eci = [1000, 2000, 3000] as EciArr3;
    const sensor = defaultSensor;
    const result = CoordinateTransforms.eci2rae(now, eci, sensor);
    expect(result).toMatchSnapshot();
  });
});
