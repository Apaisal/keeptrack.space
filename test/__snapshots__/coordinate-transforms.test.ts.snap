// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoordinateTransforms_class test_sat2ric_valid_input 1`] = `
{
  "position": Float32Array [
    7483.31494140625,
    0,
    0.00011920928955078125,
  ],
  "velocity": Float32Array [
    0,
    0,
    0,
  ],
}
`;
